autogen/coding/embedded_ipython_code_executor.py

```diff
@@ -103,35 +98,50 @@ def add_to_agent(self, agent: LLMAgent) -> None:
             Args:
                 agent (LLMAgent): The agent to add the capability to.
             """
-            agent.update_system_message(agent.system_message + self.system_message_update)
-
-    @field_validator("output_dir")
-    @classmethod
-    def _output_dir_must_exist(cls, value: str) -> str:
-        if not os.path.exists(value):
-            raise ValueError(f"Output directory {value} does not exist.")
-        return value
-
-    def __init__(self, **kwargs: Any):
-        super().__init__(**kwargs)
-        # Check if the kernel is installed.
-        if self.kernel_name not in KernelSpecManager().find_kernel_specs():
-            raise ValueError(
-                f"Kernel {self.kernel_name} is not installed. "
-                "Please first install it with "
-                f"`python -m ipykernel install --user --name {self.kernel_name}`."
-            )
-        self._kernel_manager = KernelManager(kernel_name=self.kernel_name)
-        self._kernel_manager.start_kernel()
-        self._kernel_client = self._kernel_manager.client()
-        self._kernel_client.start_channels()
-        self._timeout = self.timeout
+            agent.update_system_message(agent.system_message + self._system_message_update)
+
+    def __init__(
+        self,
+        jupyter_server: Union[JupyterConnectable, JupyterConnectionInfo],
+        kernel_name: str = "python3",
+        timeout: int = 60,
+        output_dir: Union[Path, str] = Path("."),
+        system_message_update: str = DEFAULT_SYSTEM_MESSAGE_UPDATE,
+    ):
+        if timeout < 1:
+            raise ValueError("Timeout must be greater than or equal to 1.")
+
+        if isinstance(output_dir, str):
+            output_dir = Path(output_dir)
+
+        if not output_dir.exists():
+            raise ValueError(f"Output directory {output_dir} does not exist.")
+
+        if jupyter_server is not None:
```

Comment:
What happens if jupyter_server is None?

Reply:
Accidentally left in optional behavior, fixing now